lib/gui/components/safe-webview/safe-webview.js

```
@@ -0,0 +1,231 @@
+      this.refs.webview.removeEventListener('did-get-response-details', this.didGetResponseDetails);
+      this.refs.webview.removeEventListener('will-navigate', this.willNavigate);
+      this.refs.webview.removeEventListener('ipc-message', this.ipcMessage);
+    }
+  }
+
+  /**
+   * @summary Start the loading timeout timer
+   */
+  didStartLoading() {
+    if (this.props.timeout) {
+      this.loadingTimeoutId = setTimeout(() => {
+        this.setState({
+          shouldLoad: false
+        });
+
+      // Milliseconds to wait before cancelling the page-load.
+      }, this.props.timeout);
+    }
+  }
+
+  /**
+   * @summary Set the element state to hidden
+   */
+  didFailLoad() {
+    this.setState({
+      shouldLoad: false
+    });
+  }
+
+  /**
+   * @summary Remove the timeout
+   */
+  didStopLoading() {
+
+    // We have loaded, remove the queued timeout action.
+    clearTimeout(this.loadingTimeoutId);
+  }
+
+  /**
+   * @summary Set the element state depending on the HTTP response code
+   * @param {Event} event - Event object
+   */
+  didGetResponseDetails(event) {
+
+    const {
+      httpResponseCode
+    } = event;
+    const HTTP_OK = 200;
+
+    if (httpResponseCode !== HTTP_OK) {
+      this.setState({
+        shouldLoad: false
+      });
+    }
+
+  }
+
+  /**
+   * @param {Event} event - event object
+   */
+  willNavigate(event) {
+    const url = new URL(event.url);
+
+    if (url.hash === '#external') {
+      this.refs.webview.stop();
+
+      url.hash = '';
+
+      if (url.protocol === 'http:' || url.protocol === 'https:') {
+        electron.shell.openExternal(url.href);
+      }
+    }
+  }
+
+  ipcMessage(event) {
+    const maybeArg = _.get(event.args, 0);
+
+    if (_.get(maybeArg, 'type') === 'error') {
+      analytics.logException(_.get(maybeArg, 'message'));
+
+    } else if (_.get(maybeArg, 'type') === 'event') {
+      analytics.logEvent(_.get(maybeArg, 'message'));
+    }
```

Can we log a debugging message if the type doesn't match, just to make sure we don't silently swallow anything.